### PR TITLE
Components

### PR DESCRIPTION
**Pull request by Christine-horrocks**
Rebase with development

### CODE REVIEW

File: .yardoc/checksums [hunk 1]
```diff
@@ -0,0 +1,49 @@
+app/jobs/application_job.rb 9a24c3e30cfb517943d8774df4695ba794bfd69d
```
Comment (cazwazacz) [inline .yardoc/checksums]: Please add this to GitIgnore

File: app/serializers/component_serializer/block_component_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module ComponentSerializer
+  class BlockComponentSerializer <BaseSerializer
```
Comment (cazwazacz) [inline app/serializers/component_serializer/block_component_serializer.rb]: Should this be? (missing space)
```ruby
class BlockComponentSerializer < BaseSerializer
```

Comment (cazwazacz) [inline app/serializers/component_serializer/block_component_serializer.rb]: Missing top level class documentation

File: app/serializers/component_serializer/block_component_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    # Initialise a block component with an array of objects.
+    #
+    # @param [Array<Hash>] an array of objects, each object is a component or atom.
+    # @param [String] contains the class type of the block, with a default of block.
+    def initialize(components, klass = 'block')
```
Comment (cazwazacz) [inline app/serializers/component_serializer/block_component_serializer.rb]: css class name?

Comment (cazwazacz) [reply app/serializers/component_serializer/block_component_serializer.rb]: Should this be a constant? To limit user's ability to do something weird

File: app/serializers/component_serializer/card_component_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+module ComponentSerializer
+  class CardComponentSerializer <BaseSerializer
```
Comment (cazwazacz) [inline app/serializers/component_serializer/card_component_serializer.rb]: same as above

File: app/serializers/component_serializer/card_component_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+    # Initialise a card component with an array of objects.
+    #
+    # @param [Array<Hash>] an array of objects, each object is a component or atom.
```
Comment (cazwazacz) [inline app/serializers/component_serializer/card_component_serializer.rb]: Example?

File: app/serializers/component_serializer/description_list_component_serializer.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+module ComponentSerializer
+  class DescriptionListComponentSerializer <BaseSerializer
+
+    # Initialise a paragraph component with an array of objects.
+    #
+    # @param [Array<Hash>] an array of objects, each containing 2 key/ value pairs. The first pair will have the key "medium-name", while the value will be wrapped in <dt> tags. The second key will be "medium-link", with the value being an array of strings to be wrapped in <dd> tags.
+
```
Comment (cazwazacz) [inline app/serializers/component_serializer/description_list_component_serializer.rb]: no space

File: app/serializers/component_serializer/heading_component_serializer.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
+module ComponentSerializer
+  class HeadingComponentSerializer < BaseComponentSerializer
+
+    # Initialise a paragraph component with a piece of text and an integer.
+    #
+    # @param [String] a string of text to be wrapped in <h> tags.
+    # @param [Integer] a number from 1 - 4 to be placed in the <h> tag.
+
```
Comment (cazwazacz) [inline app/serializers/component_serializer/heading_component_serializer.rb]: same as above

File: app/serializers/component_serializer/heading_component_serializer.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
+module ComponentSerializer
+  class HeadingComponentSerializer < BaseComponentSerializer
+
+    # Initialise a paragraph component with a piece of text and an integer.
```
Comment (cazwazacz) [inline app/serializers/component_serializer/heading_component_serializer.rb]: content vs test

File: app/serializers/component_serializer/heading_component_serializer.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      "heading"
+    end
+
+    def data
+      {
+        "heading": @heading,
+        "weight": @weight
+      }
+    end
+
```
Comment (cazwazacz) [inline app/serializers/component_serializer/heading_component_serializer.rb]: no space

File: app/serializers/component_serializer/heading_component_serializer.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    # @param [String] a string of text to be wrapped in <h> tags.
+    # @param [Integer] a number from 1 - 4 to be placed in the <h> tag.
+
+    def initialize(heading, weight)
+      @heading = heading
+      @weight = weight
+    end
+
+    private
+
```
Comment (cazwazacz) [inline app/serializers/component_serializer/heading_component_serializer.rb]: no space? see rubocop

File: app/serializers/component_serializer/image_component_serializer.rb [hunk 11]
```diff
@@ -1,23 +1,27 @@
...
-    def initialize(object)
-      @object = object
+
+    def initialize(figure_url, image_srcset1, image_srcset2, image_src, image_alt)
+      @figure_url = figure_url
+      @image_srcset1 = image_srcset1
+      @image_srcset2 = image_srcset2
+      @image_src = image_src
+      @image_alt = image_alt
     end
```
Comment (cazwazacz) [inline app/serializers/component_serializer/image_component_serializer.rb]:
```ruby
      @figure_url = figure_url

      @image_src = image_src
      @image_alt = image_alt

      @image_srcset1 = image_srcset1
      @image_srcset2 = image_srcset2
```

File: app/serializers/component_serializer/list_component_serializer.rb [hunk 12]
```diff
@@ -1,5 +1,10 @@
 module ComponentSerializer
-  class ListComponentSerializer < BaseSerializer
+  class ListComponentSerializer <BaseSerializer
+
+    # Initialise a list component with an array of objects.
+    #
+    # @param [String] contains the heading of the list.
+    # @param [Array<Hash>] an array of objects, each object is a component or atom.
```
Comment (cazwazacz) [inline app/serializers/component_serializer/list_component_serializer.rb]: example?

File: app/serializers/component_serializer/list_component_serializer.rb [hunk 13]
```diff
@@ -1,5 +1,10 @@
 module ComponentSerializer
-  class ListComponentSerializer < BaseSerializer
+  class ListComponentSerializer <BaseSerializer
```
Comment (cazwazacz) [inline app/serializers/component_serializer/list_component_serializer.rb]: space?

File: app/serializers/component_serializer/paragraph_component_serializer.rb [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    def name
+      "paragraph"
+    end
+
+    def data
+      {
+        "text": @text
+      }
+    end
+
```
Comment (cazwazacz) [inline app/serializers/component_serializer/paragraph_component_serializer.rb]: spaces

File: app/serializers/component_serializer/roles_list_component_serializer.rb [hunk 15]
```diff
@@ -0,0 +1,30 @@
+module ComponentSerializer
+  class RolesListComponentSerializer < BaseComponentSerializer
+    UL = 'ul'.freeze
```
Comment (cazwazacz) [inline app/serializers/component_serializer/roles_list_component_serializer.rb]: Move these to list component

File: app/serializers/component_serializer/roles_list_component_serializer.rb [hunk 16]
```diff
@@ -0,0 +1,30 @@
+module ComponentSerializer
+  class RolesListComponentSerializer < BaseComponentSerializer
```
Comment (cazwazacz) [inline app/serializers/component_serializer/roles_list_component_serializer.rb]: This should inherit from list?

File: app/serializers/component_serializer/roles_list_component_serializer.rb [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    UL = 'ul'.freeze
+    OL = 'ol'.freeze
+
+    def initialize(current_roles, type='role', list_type=ComponentSerializer::RolesListComponentSerializer::UL, class_type = 'list--pipe')
```
Comment (cazwazacz) [inline app/serializers/component_serializer/roles_list_component_serializer.rb]:
```ruby
def initialize(current_roles, type='role', list_type=ComponentSerializer::ListComponentSerializer::UL, class_type = 'pipe')
```

File: app/serializers/component_serializer/section_primary_component_serializer.rb [hunk 18]
```diff
@@ -0,0 +1,25 @@
+module ComponentSerializer
+  class SectionPrimaryComponentSerializer <BaseSerializer
+
+    # Initialise a section primary component with an array of objects.
+    #
+    # @param [Array<Hash>] an array of objects, each object is a component or atom.
+
```
Comment (cazwazacz) [inline app/serializers/component_serializer/section_primary_component_serializer.rb]: space?

File: app/serializers/page_serializer/person_show_page_serializer.rb [hunk 19]
```diff
@@ -7,25 +7,115 @@ def initialize(person, seat_incumbencies=[], committee_memberships=[], governmen
...
+  def contact_components
+    c = []
+    c << ComponentSerializer::HeadingComponentSerializer.new(t('.contact_points.contact_caps'), 2).to_h
+    c << ComponentSerializer::DescriptionListComponentSerializer.new(contact_links).to_h if @person.current_seat_incumbency.contact_points != []
+    c
+  end
+
+  def contact_links
+    contact_points = []
+   @person.try(&:current_seat_incumbency).try(&:contact_points).each do |contact_point|
```
Comment (cazwazacz) [inline app/serializers/page_serializer/person_show_page_serializer.rb]: Error prone logic

File: app/serializers/page_serializer/person_show_page_serializer.rb [hunk 20]
```diff
@@ -7,25 +7,115 @@ def initialize(person, seat_incumbencies=[], committee_memberships=[], governmen
...
 
   def title
     "#{@person.display_name} #{t('.uk_parliament')}"
   end
 
+  def section_primary_components
+    c = []
+    c << ComponentSerializer::HeadingComponentSerializer.new(@person.full_name, 1).to_h
+    c << ComponentSerializer::ParagraphComponentSerializer.new(mp_constituency_information).to_h
+    c << ComponentSerializer::ImageComponentSerializer.new(figure_url, image_srcset1, image_srcset2, image_src, image_alt).to_h if @person.image_id && @person.image_id != 'placeholder'
```
Comment (cazwazacz) [inline app/serializers/page_serializer/person_show_page_serializer.rb]: Maybe update to use srcset array?

 ```ruby
def image_srcset
    [
      generate_srcset(@person.image_id, 'CU_5:2', 732, 1464),
      generate_srcset(@person.image_id, 'MCU_3:2', 444, 888)
    ]
  end

  def generate_srcset(id, crop, small_width, large_width)
    small_src = generate_image_src(id, crop, small_width)
    large_src = generate_image_src(id, crop, large_width)
    
    "#{small_src}, #{large_src} 2x"
  end

  def generate_image_src(id, crop, width)
    "#{ENV['IMAGE_SERVICE_URL']}/#{id}.jpeg?crop=#{crop}&width=#{width}&quality=80"
  end
```

File: app/serializers/page_serializer/person_show_page_serializer.rb [hunk 21]
```diff
@@ -7,25 +7,115 @@ def initialize(person, seat_incumbencies=[], committee_memberships=[], governmen
...
+  end
+
+  def current_roles
+    history = @role_helper.create_role_history(@seat_incumbencies, @committee_memberships, @government_incumbencies, @opposition_incumbencies)
+    @role_helper.organise_roles(history[:current]) if history[:current]
+  end
+
+  def related_links_components
+    c = []
+    c << ComponentSerializer::HeadingComponentSerializer.new(t('.people.related_links.title'), 2).to_h if @person.weblinks? || (@person.image_id && @person.image_id != 'placeholder')
```
Comment (cazwazacz) [inline app/serializers/page_serializer/person_show_page_serializer.rb]: could this use a constant?

Comment (cazwazacz) [reply app/serializers/page_serializer/person_show_page_serializer.rb]: `ComponentSerializer::HeadingComponentSerializer::H2`

```ruby
ComponentSerializer::HeadingComponentSerializer.new('My Title', size: ComponentSerializer::HeadingComponentSerializer::H2).to_h 
```

Comment (cazwazacz) [reply app/serializers/page_serializer/person_show_page_serializer.rb]:
```ruby
ComponentSerializer::HeadingComponentSerializer.new('My Title', size: 2).to_h 
```

File: app/serializers/page_serializer/person_show_page_serializer.rb (hunk 19 above)
Comment (cazwazacz) [reply app/serializers/page_serializer/person_show_page_serializer.rb]: look at decorator default values

**Review by cazwazacz — CHANGES_REQUESTED** [review]

File: app/serializers/component_serializer/heading_component_serializer.rb (hunk 10 above)
Comment (Christine-horrocks) [reply app/serializers/component_serializer/heading_component_serializer.rb]: Spaced both above and below according to the style guide

File: app/serializers/component_serializer/paragraph_component_serializer.rb (hunk 14 above)
Comment (Christine-horrocks) [reply app/serializers/component_serializer/paragraph_component_serializer.rb]: The space between the end for the final method and the module and class ends is not in the style guide; however, is does make it much easier to read at a glance. I would ask to keep it.

File: app/serializers/component_serializer/roles_list_component_serializer.rb (hunk 15 above)
Comment (Christine-horrocks) [reply app/serializers/component_serializer/roles_list_component_serializer.rb]: Roles list component is extinct

File: app/serializers/component_serializer/roles_list_component_serializer.rb (hunk 16 above)
Comment (Christine-horrocks) [reply app/serializers/component_serializer/roles_list_component_serializer.rb]: see above

File: app/serializers/page_serializer/person_show_page_serializer.rb (hunk 20 above)
Comment (Christine-horrocks) [reply app/serializers/page_serializer/person_show_page_serializer.rb]: Changed the whole thing into a hash

**Review by cazwazacz — APPROVED** [review]

**Pull request merged**